File: 2020/day10.py
data = ""
with open("inputs/day10.txt", "r") as f:
    data = f.read()

adapters = [0] + sorted([int(s) for s in data.split("\n") if len(s) > 0])
adapters.append(adapters[-1] + 3)

diffs: dict[int, int] = {}
for i, a in enumerate(adapters[:-1]):
    diff = adapters[i + 1] - a
    if diff not in diffs:
        diffs[diff] = 0
    diffs[diff] += 1


def Part1() -> None:
    print("1s*3s: {}".format(diffs[1] * diffs[3]))


def Part2() -> None:
    vals: dict[int, int] = {adapters[-1]: 1}
    for a in reversed(adapters[:-1]):
        vals[a] = vals.get(a + 1, 0) + vals.get(a + 2, 0) + vals.get(a + 3, 0)
    print("Distinct Ways: {}".format(vals.get(0)))


def main() -> None:
    print("Day 10")
    Part1()
    Part2()


if __name__ == "__main__":
    main()

File: 2015/day10.ts
import { readFileSync } from "fs";

const data = readFileSync("./inputs/day10.txt", "utf-8");

function lookAndSay(input: string): string {
  let count = 1;
  let last = input.at(0);
  let output = "";
  for (const c of input.substring(1)) {
    if (c == last) {
      count++;
      continue;
    }
    output = `${output}${count}${last}`;
    last = c;
    count = 1;
  }
  output = `${output}${count}${last}`;
  return output;
}

function part1() {
  let last = data;
  for (let i = 0; i < 40; i++) {
    last = lookAndSay(last);
  }
  console.log(`40*LookAndSay: ${last.length}`);
}

function part2() {
  let last = data;
  for (let i = 0; i < 50; i++) {
    last = lookAndSay(last);
  }
  console.log(`50*LookAndSay: ${last.length}`);
}

console.log("DAY 10");
part1();
part2();
